pints/_abc/__init__.py

```diff
@@ -0,0 +1,362 @@
+        posterior.
+        """
+        return self._n_target
+
+    def parallel(self):
+        """
+        Returns the number of parallel worker processes this routine will be
+        run on, or ``False`` if parallelisation is disabled.
+        """
+        return self._n_workers if self._parallel else False
+
+    def run(self):
+        """
+        Runs the ABC sampler.
+        """
+        # Check stopping criteria
+        has_stopping_criterion = False
+        has_stopping_criterion |= (self._max_iterations is not None)
+        if not has_stopping_criterion:
+            raise ValueError('At least one stopping criterion must be set.')
+
+        # Iteration and evaluation counting
+        iteration = 0
+        evaluations = 0
+        accepted_count = 0
+
+        # Choose method to evaluate
+        f = self._error_measure
+
+        # Create evaluator object
+        if self._parallel:
+            # Use at most n_workers workers
+            n_workers = self._n_workers
+            evaluator = pints.ParallelEvaluator(f, n_workers=n_workers)
+        else:
+            evaluator = pints.SequentialEvaluator(f)
+
+        # Set up progress reporting
+        next_message = 0
+
+        # Start logging
+        logging = self._log_to_screen or self._log_filename
+        if logging:
+            if self._log_to_screen:
+                print('Using ' + str(self._sampler.name()))
+                if self._parallel:
+                    print('Running in parallel with ' + str(n_workers) +
+                          ' worker processess.')
+                else:
+                    print('Running in sequential mode.')
+
+            # Set up logger
+            logger = pints.Logger()
+            if not self._log_to_screen:
+                logger.set_stream(None)
+            if self._log_filename:
+                logger.set_filename(self._log_filename, csv=self._log_csv)
+
+            # Add fields to log
+            max_iter_guess = max(self._max_iterations or 0, 10000)
+            max_eval_guess = max_iter_guess
+            logger.add_counter('Iter.', max_value=max_iter_guess)
+            logger.add_counter('Eval.', max_value=max_eval_guess)
+            logger.add_float('Acceptance rate')
+            self._sampler._log_init(logger)
+            logger.add_time('Time m:s')
+
+        # Start sampling
+        timer = pints.Timer()
+        running = True
+
+        samples = []
+        while running:
+            iteration += 1
+
+            # Sample until a sample has been accepted
+            accepted_vals = None
+            while accepted_vals is None:
+                xs = self._sampler.ask(self._n_workers)
+                fxs = evaluator.evaluate(xs)
+                evaluations += self._n_workers
+                accepted_vals = self._sampler.tell(fxs)
+
+            # Store the accepted samples
+            accepted_count += len(accepted_vals)
+            for val in accepted_vals:
```

Comment:
Depending on the exact types used, this could a one line `samples.extend(accepted_vals)`